### ISSUE TITLE
docs: rollover

### ISSUE THREAD

**Issue by jtagcat** [overnode-org/overnode#25]
cc @avkonst 

https://overnode.org/docs/rollover-upgrade

Is there a possible delay (healthy for at least 2 hours)?

Is there a way to interrupt it, say if the health check does not exist, or did not capture an error, but the admin sees it clearly (preventing seeing your infra fall apart, without having a way in preventing it).

Docs: What happens if the rollover fails? Does it just stop, leaving say one instance in an upgraded unhealthy state?

Some images, such as [Samba](https://github.com/dperson/samba) have health checks built in. Can you override (disable) them? I'm also assuming that by default, they do get accounted for.

If the rollover `--delay` is `7200` (2 hours), then it's probably best to exit with 0 before the rollover is completed. Though with usage in scripts, a possibility to wait for the completion should be there (cron: attempt a rollover, if exit 1, webhook the admin).

And if the command exits with 0 before completion, something like `rollover status` would be very nice. The status could be monitored by the forementioned script, removing the need for a 'wait for completion' and also provides some overview for the admin who was notified via the webhook.

**Comment by avkonst** [comment]
> Is there a possible delay (healthy for at least 2 hours)?

Your health check process can do it. This relies on the docker feature: https://docs.docker.com/engine/reference/builder/#healthcheck

> Is there a way to interrupt it, say if the health check does not exist, or did not capture an error, but the admin sees it clearly (preventing seeing your infra fall apart, without having a way in preventing it).

Yes, just press Ctrl+C. It will interrupt the process.

> What happens if the rollover fails? Does it just stop, leaving say one instance in an upgraded unhealthy state?

Yes, it stops in partially upgraded state. The expectation is that the distributed system is still functional because it is usually the first instance breaks things. It gives time to understand the cause and take the required actions. One of the default actions could be to restore a configuration to a previous version and do `up` again. Or can be different action depending on a fault. At least we use this procedure. 
Makes sense?
PS: I will extend the documentation on these questions.

> Some images, such as Samba have health checks built in. Can you override (disable) them? I'm also assuming that by default, they do get accounted for.

I think you can override them. I believe it is how the Docker / Docker Compose actually works - you can override any image setting during container launch / creation. And yes, the health checks defined in an image are taken into account.


> If the rollover --delay is 7200 (2 hours), then it's probably best to exit with 0 before the rollover is completed. Though with usage in scripts, a possibility to wait for the completion should be there (cron: attempt a rollover, if exit 1, webhook the admin).
And if the command exits with 0 before completion, something like rollover status would be very nice. The status could be monitored by the forementioned script, removing the need for a 'wait for completion' and also provides some overview for the admin who was notified via the webhook.

Do you mean the case if rollover takes a lot of time? I guess you can do `overnode pull` to speed up the process. And after `overnode up --rollover` - the time will be proportional to a number of containers multiplied by health check status, which is typically should be within milliseconds or in worst case seconds. Still too long? I have not experienced much of an issue with waiting for the console output for a few minutes. But maybe larger systems might have such an issue. I guess if there is enough need for this, it can be put into a background process, which emits some status info. Makes sense?

Hope it answers your questions..

**Comment by jtagcat** [comment]
> Yes, it stops in partially upgraded state. The expectation is that the distributed system is still functional because it is usually the first instance breaks things. It gives time to understand the cause and take the required actions. One of the default actions could be to restore a configuration to a previous version and do up again. Or can be different action depending on a fault. At least we use this procedure.
> Makes sense?
https://blog.linuxserver.io/2019/10/01/updating-and-backing-up-docker-containers-with-version-control/
almost, well yes.. I think it would be very nice if ^ was built-in. 'roll**back**'

also, can you nuke the affected containers and keep the system overall running?

> the last one about 'speeding up'

The point wasn't getting it faster, on an occasion that I have a health check, that is sleep 7200, intentionally slowing it down (to see if anything is wrong on the user side). When I actually get to using it myself, I'll check back if needed.

**Comment by avkonst** [comment]
> almost, well yes.. I think it would be very nice if ^ was built-in. 'rollback'

I guess this needs to be separated into it's own ticket. Most frequently asked feature I think. But not easy to do. Maybe version 2 of the tool could do it, once the feedback for this like this is accumulated.

**Comment by avkonst** [comment]
> also, can you nuke the affected containers and keep the system overall running?

not sure... need more details on what the "overall system" is.... and how to identify the "affected containers"...

**Comment by jtagcat** [comment]
Talking of version 2 (ffs github there is never a place to talk about the project in general), how are you with languages? Most of my things have been in bash and in those cases I have felt it restricting. I was quite surprised that this thing is all shell, knowing it, my first reaction was 'no way it is _all_ in shell'. I'm probably starting to move on to the go train myself, what about you?

**Comment by jtagcat** [comment]
> > also, can you nuke the affected containers and keep the system overall running?
> 
> not sure... need more details on what the "overall system" is.... and how to identify the "affected containers"...

well I meant a workflow of:
while we have a new version available; do
we have 4 nodes
starting rollover on the first node
oh no it is bad, abort
ok, now with 3 nodes, 1 node offline (or are we with 4 nodes and when an user hits node 1 they are reaching the unhealthy instance? placing a node offline is what I mean by nuking)
let's rollback (maybe even automatically, if the rollover fails)
now with 4 nodes, 1 has been 'downgraded' (actually just 'reimaged')
let's figure out what the heck went wrong, hopefully fix it
done

**Comment by avkonst** [comment]
OK. I see what you mean. 3 related things to consider there.

1. Usually exclusion of a node happens at the level of HAProxy, which automatically detects faulty backend instances and stops sending traffic there.

2. An upgrade can happen not node-by-node but layer-by-layer. For example: 

4 nodes, a haproxy and a backend servers on each node.

`sudo overnode up --rollover backend haproxy` will upgrade in the following order: 

node1 backend
node2 backend
node3 backend
node4 backend
node1 haproxy
node2 haproxy
node3 haproxy
node4 haproxy

So a node can be only partially broken.

3. It is possible that an upgrade of a container on node 1 causes a crash of another peer layer container on node 2 or a crash of an entire upper layer. It usually means there was no enough testing done before an upgrade, but I have seen it. Currently Overnode health checks all containers on a node where it performed the last action. It can be improved to do cluster wide health checking, but it will slow things down. Do you think this feature is needed? Shall it be separated to a dedicated ticket?

**Comment by jtagcat** [comment]
> Usually exclusion of a node happens at the level of HAProxy, which automatically detects faulty backend instances and stops sending traffic there.

Err.. I'd strive for DNS level HA*. Client picks random one, when a node goes down, client picks an another one (you have multiple DNS entries with the target being different nodes).

You have to host HAProxy somewhere as well. What if that part fails?

**Comment by jtagcat** [comment]
> node1 backend
> node2 backend
> node3 backend
> node4 backend
> node1 haproxy
> node2 haproxy
> node3 haproxy
> node4 haproxy

Shouldn't it be done node1, node1, node2, node2? This way you don't have all nodes partially broken (if the combination/full upgrade doesn't go well).

**Comment by avkonst** [comment]
> Shouldn't it be done node1, node1, node2, node2? This way you don't have all nodes partially broken (if the combination/full upgrade doesn't go well).

If you omit service arguments for the up command, it will do as you said.

**Comment by avkonst** [comment]
The distributed systems are updated a layer by layer, where each layer is updated node by node. This is typical architecture.

**Comment by avkonst** [comment]
> Err.. I'd strive for DNS level HA*. Client picks random one, when a node goes down, client picks an another one (you have multiple DNS entries with the target being different nodes).

Yes, you can do it. Your front facing servers are HAProxy instances with the same DNS resolved to multiple IP addresses.

> You have to host HAProxy somewhere as well.

You host it in the same cluster. Checkout weavescope stack example. It does it.

> What if that part fails?

That is why the update is done a layer by layer.
